packages/compiler/server/serverlib.ts

```diff
@@ -0,0 +1,698 @@
+      let pkg: any;
+      const pkgPath = joinPaths(dir, "package.json");
+      const cached = fileSystemCache.get(pkgPath)?.data;
+
+      if (cached) {
+        pkg = cached;
+      } else {
+        [pkg] = await loadFile(
+          compilerHost,
+          pkgPath,
+          JSON.parse,
+          logMainFileSearchDiagnostic,
+          options
+        );
+        fileSystemCache.get(pkgPath)!.data = pkg ?? {};
+      }
+
+      if (typeof pkg?.cadlMain === "string") {
+        mainFile = pkg.cadlMain;
+      }
+
+      const candidate = joinPaths(dir, mainFile);
+      const stat = await doIO(
+        () => compilerHost.stat(candidate),
+        candidate,
+        logMainFileSearchDiagnostic,
+        options
+      );
+
+      if (stat?.isFile()) {
+        return candidate;
+      }
+
+      dir = getDirectoryPath(dir);
+    }
+
+    return path;
+
+    function logMainFileSearchDiagnostic(diagnostic: CadlDiagnostic) {
+      log(
+        `Unexpected diagnostic while looking for main file of ${path}`,
+        formatDiagnostic(diagnostic)
+      );
+    }
+  }
+
+  function inWorkspace(path: string) {
+    return workspaceFolders.some((f) => path.startsWith(f.path));
+  }
+
+  function getPath(document: TextDocument | TextDocumentIdentifier) {
+    if (isUntitled(document.uri)) {
+      return document.uri;
+    }
+    const path = resolvePath(fileURLToPath(document.uri));
+    pathToURLMap.set(path, document.uri);
+    return path;
+  }
+
+  function getURL(path: string) {
+    if (isUntitled(path)) {
+      return path;
+    }
+    return pathToURLMap.get(path) ?? pathToFileURL(path).href;
+  }
+
+  function isUntitled(pathOrUrl: string) {
+    return pathOrUrl.startsWith("untitled:");
+  }
+
+  function getDocument(path: string) {
+    const url = getURL(path);
+    return url ? host.getDocumentByURL(url) : undefined;
+  }
+
+  async function readFile(path: string): Promise<ServerSourceFile> {
+    // Try open files sent from client over LSP
+    const document = getDocument(path);
+    if (document) {
+      return {
+        document,
+        ...createSourceFile(document.getText(), path),
+      };
+    }
+
+    // Try file system cache
+    const cached = fileSystemCache.get(path);
+    if (cached) {
+      if (cached.type === "error") {
+        throw cached.error;
+      }
+      return cached.file;
+    }
+
+    // Hit the disk and cache
+    try {
+      const file = await host.compilerHost.readFile(path);
+      fileSystemCache.set(path, { type: "file", file });
+      return file;
+    } catch (error) {
+      fileSystemCache.set(path, { type: "error", error });
+      throw error;
+    }
+  }
+
+  async function stat(path: string): Promise<{ isDirectory(): boolean; isFile(): boolean }> {
+    // if we have an open document for the path or a cache entry, then we know
+    // it's a file and not a directory and needn't hit the disk.
+    if (getDocument(path) || fileSystemCache.has(path)) {
```

Comment:
There was a bug here, I fixed it in the emitter PR but maybe this could be the good place to do so as well https://github.com/microsoft/cadl/pull/41/files#diff-97b4206fa6eac301190b0164548985edc255b40970d3a46961e9a9cb83a41f83R687